### ISSUE TITLE
Negative BCE loss

### ISSUE THREAD

**Issue by Perenn1s** [milesial/Pytorch-UNet#176]
I am not sure if I were right, but maybe the negative output problem is caused by loss function
in this part:
```
    if net.n_classes > 1:
        criterion = nn.CrossEntropyLoss()
    else:
        criterion = nn.BCEWithLogitsLoss()
```
The loss function is BCEWithLogitsLoss
In my medical segmentation task, the mask is 0 or 1
So as the input goes through sigmoid, it is forced to negative in order to make output close to zeros

**Comment by milesial** [comment]
How is it a problem? Why are you looking at the raw logits before sigmoid?

**Comment by ImMial** [comment]
> 
> 
> How is it a problem? Why are you looking at the raw logits before sigmoid?


My problem is having negative output in another dataset.
I use another dataset and modified the dataloader, most pixels in the gt mask is 0 and some are 1. Then after some optimization process, outputs are fixed in negative values. I suppose it should be caused by sigmoid inside BCEWithLogitsLoss: only negative input will yield value near zero, so it backwards the network to negative outputs
I know it may violate the U-Net architecture, I add a sigmoid in the lastconv output, so it seems to be functioning

**Comment by milesial** [comment]
You're supposed to apply a sigmoid, how else would you get probabilities? That's what the predict script does

**Comment by RaindyLihuahua** [comment]
hey, I have the same question, and I train unet with my medical dataset too. The BCEWithLogitsLoss also have negative output. Even I add a sigmoid to lastconv and use BCEloss, it also output negative. I spend two days to work it out and now I still don't know why.

**Comment by milesial** [comment]
@RaindyLihuahua The BCE loss can't be negative so it just means you are loading your masks wrong. Check that your true masks match the specs.

**Comment by RaindyLihuahua** [comment]
> @RaindyLihuahua The BCE loss can't be negative so it just means you are loading your masks wrong. Check that your true masks match the specs.

Yes, it is. Something wrong at `BasicDataset`
`      if img_trans.max() > 1:
            img_trans = img_trans / 255`
this if sometimes don't work and img_trans.max() is 255, it makes BCEloss negative. I don't know why, it maybe a bug. I delete this line and the code runs good. Anyway, my question is solved, thank you!

**Comment by milesial** [comment]
That's really strange, I'll leave this open for others to investigate. The code is there so the images should be transformed.

**Comment by milesial** [comment]
Closing, as this issue is now outdated after the recent code refactor.
